### ISSUE TITLE
test-suite failure on Debian 9: cppd      FAIL     0.01 s (killed by signal 11 SIGSEGV)

### ISSUE THREAD

**Issue by mbiebl** [mesonbuild/meson#4049]
Trying to run the test-suite on Debian 9, the following test fails:
```
2/2 cppd                                    FAIL     0.01 s (killed by signal 11 SIGSEGV)
```
might be related to
```
Code:
 
        #include <linux/if.h>
Compiler stdout:
 
Compiler stderr:
 In file included from /tmp/tmpp25lvr20/testfile.c:2:0:
/usr/include/linux/if.h:234:19: error: field ‘ifru_addr’ has incomplete type
   struct sockaddr ifru_addr;
                   ^~~~~~~~~
/usr/include/linux/if.h:235:19: error: field ‘ifru_dstaddr’ has incomplete type
   struct sockaddr ifru_dstaddr;
                   ^~~~~~~~~~~~
/usr/include/linux/if.h:236:19: error: field ‘ifru_broadaddr’ has incomplete type
   struct sockaddr ifru_broadaddr;
                   ^~~~~~~~~~~~~~
/usr/include/linux/if.h:237:19: error: field ‘ifru_netmask’ has incomplete type
   struct sockaddr ifru_netmask;
                   ^~~~~~~~~~~~
/usr/include/linux/if.h:238:20: error: field ‘ifru_hwaddr’ has incomplete type
   struct  sockaddr ifru_hwaddr;
                    ^~~~~~~~~~~

Check usable header "linux/if.h": NO

meson.build:36:4: ERROR:  Assert failed: Could not find <linux/if.h>
```
The following versions are installed:
```
gcc:
  Installed: 4:6.3.0-4
  Candidate: 4:6.3.0-4
  Version table:
 *** 4:6.3.0-4 500
        500 http://ftp.de.debian.org/debian stretch/main amd64 Packages
        100 /var/lib/dpkg/status
libc6-dev:
  Installed: 2.24-11+deb9u3
  Candidate: 2.24-11+deb9u3
  Version table:
 *** 2.24-11+deb9u3 500
        500 http://ftp.de.debian.org/debian stretch/main amd64 Packages
        100 /var/lib/dpkg/status
linux-libc-dev:
  Installed: 4.9.110-1
  Candidate: 4.9.110-1
  Version table:
 *** 4.9.110-1 500
        500 http://ftp.de.debian.org/debian stretch/main amd64 Packages
        100 /var/lib/dpkg/status
```

**Comment by mbiebl** [comment]
Full build log 
[meson-0.47.1-backports.txt](https://github.com/mesonbuild/meson/files/2299664/meson-0.47.1-backports.txt)

**Comment by nirbheek** [comment]
> meson.build:36:4: ERROR: Assert failed: Could not find <linux/if.h>

This is unrelated to the `cppd` test. You just don't have the linux-headers package installed, I suspect. That package is not required by Meson, but is used by the tests to check that `cc.check_header()` works as expected.

**Comment by nirbheek** [comment]
Do you think you can run the `d/10 d cpp` test by hand and run the tests under gdb to get a backtrace?

**Comment by mbiebl** [comment]
I have no experience with the D language. Can you give some instructions how I can run the test by hand?

**Comment by mbiebl** [comment]
> meson.build:36:4: ERROR: Assert failed: Could not find <linux/if.h>
> 
> This is unrelated to the cppd test. You just don't have the linux-headers package installed, I suspect. That package is not required by Meson, but is used by the tests (on Linux) to check that cc.check_header() works as expected.

Installed the linux-kernel-headers package but this issue was not fixed. Searching the internet, it is suggested to include <sys/socket.h> before <linux/if.h>. If this header is problematic, why not check a header directly from libc?

**Comment by mbiebl** [comment]
Split off the <linux/if.h> issue into a separate bug report #4053

**Comment by nirbheek** [comment]
> I have no experience with the D language. Can you give some instructions how I can run the test by hand?

You don't need to do anything special. Just run `meson _build && ninja -C _build && ./_build/cppd`

Then you can run that under gdb too, although I am not sure how well gdb will react to mixed C++ and D executables.

**Comment by mbiebl** [comment]
```
ninja: Entering directory `_build/'
[1/6] Compiling C++ object 'cppd@exe/cppmain.cpp.o'.
[2/6] Compiling D object 'dcpp@exe/dmain.d.o'.
[3/6] Compiling C++ object 'dcpp@exe/libfile.cpp.o'.
[4/6] Linking target dcpp.
FAILED: dcpp»
gdc  -o dcpp 'dcpp@exe/dmain.d.o' 'dcpp@exe/libfile.cpp.o'»»»
dcpp@exe/libfile.cpp.o: In function `print_hello(int)':
/tmp/meson-0.47.1/test cases/d/10 d cpp/_build/../libfile.cpp:4: undefined reference to `std::cout'
/tmp/meson-0.47.1/test cases/d/10 d cpp/_build/../libfile.cpp:4: undefined reference to `std::basic_ostream<char, std::char_traits<char> >& std::operator<< <std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*)'
/tmp/meson-0.47.1/test cases/d/10 d cpp/_build/../libfile.cpp:4: undefined reference to `std::ostream::operator<<(int)'
/tmp/meson-0.47.1/test cases/d/10 d cpp/_build/../libfile.cpp:4: undefined reference to `std::basic_ostream<char, std::char_traits<char> >& std::operator<< <std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*)'
dcpp@exe/libfile.cpp.o: In function `__static_initialization_and_destruction_0(int, int)':
/usr/include/c++/6/iostream:74: undefined reference to `std::ios_base::Init::Init()'
/usr/include/c++/6/iostream:74: undefined reference to `std::ios_base::Init::~Init()'
collect2: error: ld returned 1 exit status
[5/6] Compiling D object 'cppd@exe/libfile.d.o'.
ninja: build stopped: subcommand failed.
```

**Comment by mbiebl** [comment]
Running this under Debian sid, I get
```
[1/6] c++ -Icppd@exe -I. -I.. -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -O0 -g  -MD -MQ 'cppd@exe/cppmain.cpp.o' -MF 'cppd@exe/cppmain.cpp.o.d' -o 'cppd@exe/cppmain.cpp.o' -c ../cppmain.cpp
[2/6] gdc -Idcpp@exe -I. -I.. -fdiagnostics-color=always -Wall -Wdeprecated -g -O0  -MD -MQ 'dcpp@exe/dmain.d.o' -MF 'dcpp@exe/dmain.d.o.deps' -o 'dcpp@exe/dmain.d.o' -c ../dmain.d
[3/6] c++ -Idcpp@exe -I. -I.. -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -O0 -g  -MD -MQ 'dcpp@exe/libfile.cpp.o' -MF 'dcpp@exe/libfile.cpp.o.d' -o 'dcpp@exe/libfile.cpp.o' -c ../libfile.cpp
[4/6] gdc  -o dcpp 'dcpp@exe/dmain.d.o' 'dcpp@exe/libfile.cpp.o' -lstdc++  
[5/6] gdc -Icppd@exe -I. -I.. -fdiagnostics-color=always -Wall -Wdeprecated -g -O0  -MD -MQ 'cppd@exe/libfile.d.o' -MF 'cppd@exe/libfile.d.o.deps' -o 'cppd@exe/libfile.d.o' -c ../libfile.d
[6/6] gdc  -o cppd 'cppd@exe/cppmain.cpp.o' 'cppd@exe/libfile.d.o' -lstdc++  
```
Note how ```-lstdc++``` is added there.

**Comment by nirbheek** [comment]
What environment did you run the failing build under? That wasn't there in the test log failure that you attached above:

```
[1/6] c++ -Icppd@exe -I. '-I../test cases/d/10 d cpp' -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -O0 -g -g -O2 -fdebug-prefix-map=/build/meson-0.47.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -MD -MQ 'cppd@exe/cppmain.cpp.o' -MF 'cppd@exe/cppmain.cpp.o.d' -o 'cppd@exe/cppmain.cpp.o' -c '../test cases/d/10 d cpp/cppmain.cpp'
[2/6] gdc -Idcpp@exe -I. '-I../test cases/d/10 d cpp' -fdiagnostics-color=always -Wall -Wdeprecated -g -O0   -o 'dcpp@exe/dmain.d.o' -c '../test cases/d/10 d cpp/dmain.d'
[3/6] c++ -Idcpp@exe -I. '-I../test cases/d/10 d cpp' -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -O0 -g -g -O2 -fdebug-prefix-map=/build/meson-0.47.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -MD -MQ 'dcpp@exe/libfile.cpp.o' -MF 'dcpp@exe/libfile.cpp.o.d' -o 'dcpp@exe/libfile.cpp.o' -c '../test cases/d/10 d cpp/libfile.cpp'
[4/6] gdc  -o dcpp 'dcpp@exe/dmain.d.o' 'dcpp@exe/libfile.cpp.o' -Wl,-z,relro -lstdc++  
[5/6] gdc -Icppd@exe -I. '-I../test cases/d/10 d cpp' -fdiagnostics-color=always -Wall -Wdeprecated -g -O0   -o 'cppd@exe/libfile.d.o' -c '../test cases/d/10 d cpp/libfile.d'
[6/6] gdc  -o cppd 'cppd@exe/cppmain.cpp.o' 'cppd@exe/libfile.d.o' -Wl,-z,relro -lstdc++  
1/2 dcpp                                    OK       0.02 s 
2/2 cppd                                    FAIL     0.01 s (killed by signal 11 SIGSEGV)
```

**Comment by mbiebl** [comment]
The environment is a Debian stretch build chroot.
The original bug report was generated by running the Debian package build (via dpkg-buildpackage).
The second, failing build was generated by installing meson, then running the commands you listed above

**Comment by mbiebl** [comment]
The installed meson version in the Debian stretch chroot is 0.45.1-2~bpo9+3.
I guess I should use meson from the build directory instead...

**Comment by nirbheek** [comment]
@GoaLitiuM does this backtrace look like anything to you?

**Comment by GoaLitiuM** [comment]
The test was changed in #3981, so try running the tests with that PR. The culprit in my case was uninitialized D runtime, which happens if the function is called from C++ main, since the D main does the runtime initialization automatically.

**Comment by nirbheek** [comment]
If this issue is fixed by #3981, we will want to cherry-pick that change into 0.47.2 so that this issue is resolved.

**Comment by mbiebl** [comment]
@nirbheek I've applied the PR on top of git master and the test now passes
